src/app/dataset-view/additional-components/lineage-component/lineage.component.ts

```diff
@@ -69,62 +56,92 @@ export class LineageComponent extends BaseComponent implements OnInit {
-                if (typeof cluster.childNodeIds === "undefined") {
-                    cluster.childNodeIds = [];
-                }
-
-                if (cluster.label === dataset.kind) {
-                    cluster.childNodeIds.push(dataset.id);
-                }
-                return cluster;
-            });
-        });
-
         const edges = lineageUpdate.edges;
         const currentDataset = lineageUpdate.origin;
-
         this.initLineageGraphProperty();
-
         this.isAvailableLineageGraph = edges.length !== 0;
-
-        const uniqueDatasets: Record<string, DatasetBasicsFragment> = {};
-        edges.forEach((edge: DatasetBasicsFragment[]) =>
-            edge.forEach((dataset: DatasetBasicsFragment) => {
+        const uniqueDatasets: Record<string, DatasetLineageBasics> = {};
+        edges.forEach((edge: DatasetLineageBasics[]) =>
+            edge.forEach((dataset: DatasetLineageBasics) => {
                 uniqueDatasets[dataset.id] = dataset;
             }),
         );
+        this.addSourceGraphNodes(Object.values(uniqueDatasets));
+        this.addDatasetGraphNodes(Object.values(uniqueDatasets), edges, currentDataset);
+    }
 
-        for (const [id, dataset] of Object.entries(uniqueDatasets)) {
+    private sanitizeID(id: string): string {
+        return id.replace(/:/g, "");
+    }
+
+    private getDomainFromUrl(url: string): string {
+        const protocols = ["http", "https", "ftp"];
+        return protocols.some((protocol) => url.includes(protocol)) ? new URL(url).hostname : url;
+    }
+
+    private addSourceGraphNodes(data: DatasetLineageBasics[]): void {
+        const extraNodes = data.filter(
+            (item: DatasetLineageBasics) =>
+                item.kind === DatasetKind.Root && item.metadata.currentSource?.fetch.__typename === "FetchStepUrl",
+        );
+        extraNodes.forEach((node: DatasetLineageBasics) => {
+            const id = this.sanitizeID(node.id);
+            const label = this.getDomainFromUrl(
+                node.metadata.currentSource?.fetch.__typename === "FetchStepUrl"
+                    ? node.metadata.currentSource.fetch.url
+                    : "Other source",
+            );
+            const indexExistingSource = this.lineageGraphNodes.findIndex((node) => node.label === label);
+            if (indexExistingSource === -1) {
+                this.lineageGraphNodes.push({
+                    id: "extra-node-" + id,
+                    label,
+                    data: {
+                        nodeKind: LineageGraphNodeType.Source,
+                        nodeDataObject: {},
+                    } as LineageGraphNodeData,
+                });
+            }
+            const validId = indexExistingSource == -1 ? id : this.lineageGraphLink[indexExistingSource].target;
+            this.lineageGraphLink.push({
+                id: `extra-node-${validId}__and__${id}`,
```

Comment:
Consider encapsulating naming of the nodes in one place, i.e., add functions to create ID for each type of node - it would know the repeating part and receive the variative part as an argument.

Reply:
Fixed